### PR TITLE
WiP: PR0 (SPI write prevention through chipset locking) for nv4x_adl, setting base for other platforms/downstream forks supporting >=Skylake+

### PR DESCRIPTION
**Pull request by tlaurion**
**PRIOR OF MERGE: only nv4x_adl as a >=Skylake platform has PR0 enabled as per this PR.** This PR enables future enablement of PR0 once downstream forks cherry-pick+adapt https://review.coreboot.org/c/coreboot/+/85278

Politics (I will let downstream coreboot forks opt-in, this is not my fight, draining) at https://github.com/linuxboot/heads/pull/1818#issuecomment-2508013113

Hows: 
- https://github.com/linuxboot/heads/blob/f8b03b30878debd3e3dfd8c8365ea245ccedc8c2/patches/coreboot-dasharo-unreleased/0002-pr0_chipset_locking-post_skylake.patch is applied as patch to the fork (note path `patches/coreboot-dasharo-unreleased`)
- https://github.com/linuxboot/heads/pull/1818/files#diff-7c47e775c4b07a59cc813789395394f1b213deaba73610d283ad88fd521afd56
- https://github.com/linuxboot/heads/pull/1818/files#diff-b12b179ed02b616928b3971aec34bcb850f0c0d773eb98b17eba91c76750cf82

Notes: ns50 is prepared on coreboot side, but board doesn't include io386 nor enables PR0 in board config:
- coreboot config prepared for enablement: https://github.com/linuxboot/heads/pull/1818/files#diff-836ff08c654dbca47c03cabc60603ed0f0d6e24228017bdd8dbaadbd5e9993e0
- board config related config commented, not enabled: https://github.com/linuxboot/heads/pull/1818/files#diff-15afe4b741eb85edd66b8661bfff49fd8b53459a47cf8ed3b91d3fea67a86f6c

For ns50 enablement in next PR, see https://github.com/linuxboot/heads/pull/1818#issuecomment-2508097578

### TLDR
**NEXT: Librems/ v540TU #1846 / V560TU #1835 :  will need to have https://review.coreboot.org/c/coreboot/+/85278 adapted patch to their coreboot fork if not coreboot upstream**

----

@miczyg1 pushed some changes in dasharo coreboot fork to enable SMM PR0:
This cherry-pick this commit https://github.com/Dasharo/coreboot/commit/ff22122c229bbe2109de92ded773493428f7ece9

To test this PR for nv41 users (EC needs to be up to date otherwise problems might occur):
- Flash this PR rom
- go to recovery shell
- call lock_chip from command line (locks PR0)
- call flash-gui.sh
- attempt to flash master's rom, see failing: expected
- reboot (cancels PR0)
- flash back master's rom from gui: see success.

At term, this PR will bring coreboot Skylake+ equivalent of <=Skylake for PR0 chipset locking of SPI WP (SPI RW protection from OS: OS cannot write to SPI, only Heads can flash internally when enabled per board config + coreboot config on tested platforms).

@MrChromebox you might take a look at https://github.com/linuxboot/heads/issues/1659, which this PR will fix when working on all Skylake+ based boards merged for <Skyylake at https://github.com/linuxboot/heads/pull/1373

---

NOTE: if for whatever reason, you would love to **disable PR0 locking on a single boot option**, you can do so through configuration settings menu under heads as screenshots show under https://github.com/linuxboot/heads/pull/1373#issuecomment-1599257105 :

> signal-2023-06-20-135059
> signal-2023-06-20-135116
> signal-2023-06-20-135128
> signal-2023-06-20-135145
> 
> On boot, the presence of "Finalizing chipset" disappears, where its present on normal boot path otherwise.

---

At term, this will fix https://github.com/linuxboot/heads/issues/1659

TODO:
- [x] nv41 functional, flashrom tested internally after calling lock_chip: unsuccessful (ok!)
- [x] fix and merge #1814 
- [ ] add patch to other forks
  - [x] or merge this PR while other forks pick it up?
- [x] adapt lock_chip and verbiage if not just fro pre-skylake in codebase
- [x] +1 review on coreboot when patchset tested and working for all skylake+
- [ ] 2024-11-29: retest nv4x_adl rebased on master per https://github.com/linuxboot/heads/pull/1818/commits/03ba3864dbb3cc7f999e658ec5a067c71985a465
- [ ] merge

--

@JonathonHall-Purism patch doesn't apply cleanly on purism fork

### CODE REVIEW

**Comment by tlaurion** [comment]
@miczyg1 false alarm, it was again my external programmer's fault. Tigard doesn't work well but again I have a leg of wson8 probe broken.

This is good to go!

**Comment by tlaurion** [comment]
Question here is what do we want prior of merging @miczyg1 @JonathonHall-Purism.

All Skylake+ platforms implementing PR0?

**Comment by tlaurion** [comment]
> Question here is what do we want prior of merging @miczyg1 @JonathonHall-Purism.
> 
> All Skylake+ platforms implementing PR0?

@macpijan @JonathonHall-Purism let me know if something else needs to be done prior of #1821 deadline. Could be in or not: I would prefer this to be in.

**Comment by tlaurion** [comment]
Rebasing on master.

**Comment by JonathonHall-Purism** [comment]
@tlaurion @miczyg1 Nice work here, I'm thrilled to see this.  I hope we can get it to work :crossed_fingers: 

What SoCs has this been tested on so far?  I applied the coreboot patch to Purism's coreboot fork and built this for the Librem 14; the OS is still able to write to the firmware.  The coreboot patch is implemented for cannonlake but I don't know if that was tested.

I did see `Finalizing chipset Write Protection through SMI PR0 lockdown call`, so the config is right and Heads did appear to try to lock flash.  (Full disclosure, I booted manually with `basic-autoboot.sh` from the recovery shell, but this uses the same `kexec-boot` path and I did see that trace.)

The OS was still able to write to flash (in this case I changed the serial number using our coreboot utility: that reads flash, uses cbfstool to change the serial, then writes flash).

I haven't looked into it any further.  What SoCs have you all tested?  I'd be happy to help get this working for CNL and other SoCs but I probably will not be able to do it prior to the feature freeze.

> Question here is what do we want prior of merging @miczyg1 @JonathonHall-Purism.
> 
> All Skylake+ platforms implementing PR0?

I'd suggest:
* Heads configures all supported platforms with this behavior (we want similar configs for all boards upstream)
* Offer a configuration toggle to control it for now - I will want to have better integration with our coreboot utility and other tooling before I enable it in PureBoot (they need to tell the user what to do, can't just fail in confusion).  (I'm happy to help with this too.)

**Comment by tlaurion** [comment]
> * Offer a configuration toggle to control it for now

It's already there as per original PR0 PR: as of now, user can turn it off for one boot under config settings toggle. 
@JonathonHall-Purism what would you suggest instead?

**Comment by tlaurion** [comment]
Rebasing on master

**Comment by miczyg1** [comment]
> @tlaurion @miczyg1 Nice work here, I'm thrilled to see this. I hope we can get it to work 🤞
> 
> What SoCs has this been tested on so far? I applied the coreboot patch to Purism's coreboot fork and built this for the Librem 14; the OS is still able to write to the firmware. The coreboot patch is implemented for cannonlake but I don't know if that was tested.
> 
> I did see `Finalizing chipset Write Protection through SMI PR0 lockdown call`, so the config is right and Heads did appear to try to lock flash. (Full disclosure, I booted manually with `basic-autoboot.sh` from the recovery shell, but this uses the same `kexec-boot` path and I did see that trace.)
> 
> The OS was still able to write to flash (in this case I changed the serial number using our coreboot utility: that reads flash, uses cbfstool to change the serial, then writes flash).
> 
> I haven't looked into it any further. What SoCs have you all tested? I'd be happy to help get this working for CNL and other SoCs but I probably will not be able to do it prior to the feature freeze.
> 
> > Question here is what do we want prior of merging @miczyg1 @JonathonHall-Purism.
> > All Skylake+ platforms implementing PR0?
> 
> I'd suggest:
> 
> * Heads configures all supported platforms with this behavior (we want similar configs for all boards upstream)
> * Offer a configuration toggle to control it for now - I will want to have better integration with our coreboot utility and other tooling before I enable it in PureBoot (they need to tell the user what to do, can't just fail in confusion).  (I'm happy to help with this too.)

I checked it on Alder Lake and Comet Lake (so using coreboot's soc/cannonlake) and it worked.

**Comment by tlaurion** [comment]
@JonathonHall-Purism ping

**Comment by tlaurion** [comment]
@miczyg1 is this linked to an upstream patchset I've missed for upstream review? Maybe this would make this go faster? (issue thought unfixable for ~3y even though under vaultboot for skylake, see https://github.com/linuxboot/heads/pull/326#issuecomment-939176320)

**Comment by miczyg1** [comment]
Pushed: https://review.coreboot.org/c/coreboot/+/85278

**Comment by tlaurion** [comment]
bumped to last changes over coreboot pointed patchset, rebased on master under latest commit.

@miczyg1 I replied to @paulmenzel at https://review.coreboot.org/c/coreboot/+/85278/comment/1bc100aa_c7700a9d/

**Comment by miczyg1** [comment]
> @tlaurion thank you for persistently pushing for upstreaming/ contributing back to main coreboot tree. The community thanks you ☺️

No worries @LeanSheng , it was supposed to go upstream from the beginning. It was just simpler for me to quickly draft a PoC on a fork and test on HW I had with me locally.

**Comment by tlaurion** [comment]
@miczyg1 Cannot use patchset against current dasharo fork used by nv41, so will revert last commit until either Dasharo bumps the Pr0 path that works for dasharo fork or bumps dasharo fork so that upsteam pathc applies cleanly here

Logs at https://app.circleci.com/pipelines/github/tlaurion/heads/3009/workflows/70ac4899-04e1-4527-883b-f6eea5070c0d/jobs/58411/parallel-runs/0/steps/0-102


@macpijan cc

**Comment by tlaurion** [comment]
Putting back to draft meanwhile

**Comment by miczyg1** [comment]
> @miczyg1 Cannot use patchset against current dasharo fork used by nv41, so will revert last commit until either Dasharo bumps the Pr0 path that works for dasharo fork or bumps dasharo fork so that upsteam pathc applies cleanly here
> 

I would focus on testing the boards using upstream coreboot in heads. When we reach the final shape of the patch on gerrit and get it merged, I can cherry-pick it to Dasharo and test the remaining boards based on Dasharo fork.

**Comment by macpijan** [comment]
nv41 is not part of upstream, if you mean to use upstream for this board

**Comment by miczyg1** [comment]
> nv41 is not part of upstream, if you mean to use upstream for this board

I meant all boards that do not use coreboot forks in heads.

**Comment by tlaurion** [comment]
> I would focus on testing the boards using upstream coreboot in heads. 

@miczyg1 (dasharo) @JonathonHall-Purism (purism) forks:
TLDR: from above comment this is not possible today because no >=skylake boards are using coreboot master today, therefore patches needs to be in forks ot be tested.


A reminder that Heads is a user of downstream/upstream coreboot forks. Nothing much I can do here.

**Comment by tlaurion** [comment]
@JonathonHall-Purism @miczyg1 As far as I'm concerned, this PR is ready to merge as it is, in a functional state until all comments above are resolved in another PR/downstream in coreboot forks.

Waiting for comments otherwise merging at end of the day.

**Comment by tlaurion** [comment]
.@nestire @daringer @jans23: Removing ns50 lock_chip because I cannot test. Simple to add, just revert next commit in a PR after testing it and I will merge.

**Comment by miczyg1** [comment]
> TLDR: from above comment this is not possible today because no >=skylake boards are using coreboot master today, therefore patches needs to be in forks ot be tested.

Ugh... I was not aware of that. Allright. Then I will focus on getting the patch merged upstream, then cherry-pick to dasharo.

**Comment by LeanSheng** [comment]
> > TLDR: from above comment this is not possible today because no >=skylake boards are using coreboot master today, therefore patches needs to be in forks ot be tested.
> 
> Ugh... I was not aware of that. Allright. Then I will focus on getting the patch merged upstream, then cherry-pick to dasharo.

@miczyg1 it has been awhile, any update on coreboot patch?

**Comment by tlaurion** [comment]
> > > TLDR: from above comment this is not possible today because no >=skylake boards are using coreboot master today, therefore patches needs to be in forks ot be tested.
> > 
> > 
> > Ugh... I was not aware of that. Allright. Then I will focus on getting the patch merged upstream, then cherry-pick to dasharo.
> 
> @miczyg1 it has been awhile, any update on coreboot patch?

@miczyg1 there is merge conflict https://review.coreboot.org/c/coreboot/+/85278

Ideally, links to downstream commits and versions of coreboot should be clarified under that merge request I guess?
Not sure how to make this merged upstream here.

**Comment by LeanSheng** [comment]
@tlaurion yes you are right. thats y we always ask people to upstream to coreboot directly and not to create another fork and keep things there. coreboot is not the same as linux kernel which has big ecosystem; but rather niche and need more support for growth, especially for companies that using coreboot for business should be uphold bigger responsibilities for upstreaming and maintaining. for many years we have been working hard to promote coreboot, have to work together.

**Comment by miczyg1** [comment]
> @miczyg1 it has been awhile, any update on coreboot patch?

I will try to find some time to push new patchset.

**Comment by LeanSheng** [comment]
@miczyg1 Merry Christmas and Happy New Year! Any update? What is the plan to update it?

**Comment by tlaurion** [comment]
This is important.

> https://eclypsium.com/blog/firmware-security-realizations-part-3-spi-write-protections

_Originally posted by @tlaurion in https://github.com/linuxboot/heads/issues/1659#issuecomment-2096933072_
